[Core] Adding to modelpart create geometries methods

loumalouomega opened the pull request:
Depends on #6281. It adds the CreateNewGeometry methods to the model part. In order to do that adds the Create method considering array of points + Id to the main geometries

loumalouomega committed:
Merge branch 'master' into core/modelpart-create-geometries-methods

loumalouomega committed:
Adding create with Id

loumalouomega committed:
Adding create with Id for Hexahedra3D27

loumalouomega committed:
Adding create with Id for Hexahedra3D27

loumalouomega committed:
Adding create with Id for Hexahedra3D8

loumalouomega committed:
Adding create with Id for Line2D2

loumalouomega committed:
Adding create with Id for Line2D3

loumalouomega committed:
Adding create with Id for Line3D2

loumalouomega committed:
Adding create with Id for Line3D3

loumalouomega committed:
Adding create with Id for Point2D

loumalouomega committed:
Adding create with Id for Point3D

loumalouomega committed:
Adding create with Id for Prism3D15

loumalouomega committed:
Adding create with Id for Prism3D6

loumalouomega committed:
Adding create with Id for QuadraturePoints

loumalouomega committed:
Adding create with Id for Quadrilateral2D4

loumalouomega committed:
Adding create with Id for Quadrilateral2D8

loumalouomega committed:
Adding create with Id for Quadrilateral2D9

loumalouomega committed:
Adding create with Id for Quadrilateral3D4

loumalouomega committed:
Adding create with Id for Quadrialteral3D8

loumalouomega committed:
Adding create with Id for Quadrilateral3D9

loumalouomega committed:
Adding create with Id for Sphere3D

loumalouomega committed:
Adding create with Id for Tetrahedra3D10

loumalouomega committed:
Adding create with Id for Tetrahedra3D4

loumalouomega committed:
Adding create with Id for Triangle2D3

loumalouomega committed:
Adding create with Id for Triangle2D6

loumalouomega committed:
Adding create with Id for Triangle3D3

loumalouomega committed:
Adding create with Id for Triangle3D6

loumalouomega committed:
Adding CreateNewGeometry

loumalouomega committed:
Exposing to python

loumalouomega committed:
Test (WIP)

loumalouomega committed:
Fix

tteschemacher committed:
Add Create with GeometryName to geometry.h

tteschemacher committed:
Update comment

tteschemacher committed:
Adding Create with GeometryName to Hexahedra3D20

tteschemacher committed:
Adding Create with GeometryName to Hexahedra3D27

tteschemacher committed:
Adding Create with GeometryName to Hexahedra3D8

tteschemacher committed:
Adding Create with GeometryName to Line2D2

tteschemacher committed:
Adding Create with GeometryName to Line2D3

tteschemacher committed:
Adding Create with GeometryName to Line3D2

tteschemacher committed:
Adding Create for GeometryName to Line3D3

tteschemacher committed:
Adding Create with GeometryName to Point2D

tteschemacher committed:
Adding Create with GeometryName to Point3D

tteschemacher committed:
Adding Create with GeometryName to Prism3D15

tteschemacher committed:
Adding Create with GeometryName to Prism3D6

tteschemacher committed:
Adding Create with GeometryName to QuadratuePointGeometry

tteschemacher committed:
Adding Create with GeometryName to Quadrilateral2D4

tteschemacher committed:
Adding Create with GeometryName to Quadrilateral2D8

tteschemacher committed:
Adding Create with GeometryName to Quadrilateral2D9

tteschemacher committed:
Adding Create with GeometryName to Quadrilateral3D4

tteschemacher committed:
Adding Create with GeometryName to Quadrilateral3D8

tteschemacher committed:
Adding Create with GeometryName to Quadrilateral3D9

tteschemacher committed:
Adding Create with GeometryName to Tetrahedra3D10

tteschemacher committed:
Adding Create with GeometryName to Tetrahedra3D4

tteschemacher committed:
Adding Create with GeometryName to Triangle2D3

tteschemacher committed:
Adding Create with GeometryName to Triangle2D6

tteschemacher committed:
Adding Create with GeometryName to Triangle3D3

tteschemacher committed:
Adding Create with GeometryName to Triangle3D6

tteschemacher committed:
fix type of GeometryName

tteschemacher committed:
Fix name from NewId to NewGeometryId

tteschemacher committed:
Change Id to GeometryId and GeometryName to GeometryTypeName

tteschemacher committed:
Fix Hexahedra3D27

tteschemacher committed:
Fixing QuadraturePointGeometry

tteschemacher committed:
Adding CreateNewGeometry with GeometryName to ModelPart

tteschemacher committed:
Adding ModelPartCreateNewGeometry with GeometryName to Python

tteschemacher committed:
Adding CreateNewGeometry to model_part without identifier/id

tteschemacher committed:
Adding CreateNewModelGeometry without id to python

tteschemacher committed:
Update test to check different create functions

tteschemacher committed:
Update hexahedra_3d_27.h

tteschemacher committed:
Update quadrilateral_3d_9.h

tteschemacher commented:
Thank you @loumalouomega.

I added the additional Creates with identifier name. Also the Create without id is possible, thus I added it to the mp.

loumalouomega commented:
> Thank you @loumalouomega.
> 
> I added the additional Creates with identifier name. Also the Create without id is possible, thus I added it to the mp.

Now the compilation fails :(

tteschemacher commented:
I'll check it

tteschemacher committed:
fix virtual to typename

tteschemacher committed:
mark overrides

tteschemacher commented:
Now it seems to work again. :)

I am actually fine with merging it. But I think we should wait for @KratosMultiphysics/technical-committee or @pooyan-dadvand, shouldn't we?

loumalouomega commented:
Remember first to merge #6281(this makes easier to review) @pooyan-dadvand @RiccardoRossi

loumalouomega committed:
Merge branch 'master' into core/modelpart-create-geometries-methods

pooyan-dadvand commented:
Is it still based on the closed PR?

loumalouomega commented:
> 
> 
> Is it still based on the closed PR?

This one is not polluted, we should merge #6367 first and solve eventual conflicts

loumalouomega committed:
Merge branch 'master' into core/modelpart-create-geometries-methods

tteschemacher committed:
make const string to const string& in CreateGeometry methods

loumalouomega committed:
Merge branch 'master' into core/modelpart-create-geometries-methods

tteschemacher committed:
Update geometry.h

loumalouomega commented:
Thanks @tteschemacher

loumalouomega commented:
This is ready to be reviwed @KratosMultiphysics/technical-committee

tteschemacher commented on kratos/geometries/geometry.h:
```diff
@@ -647,11 +647,46 @@ class Geometry
+     */
+    virtual Pointer Create(
+        IndexType NewGeometryId,
+        PointsArrayType const& ThisPoints
```
Can we use here Geometry instead of the PointsArrayType? 

That would allow to use the new geometries which have additional data enhenced.

loumalouomega replied:
Sure

loumalouomega replied:
@tteschemacher done

loumalouomega committed:
Merge branch 'master' into core/modelpart-create-geometries-methods

loumalouomega committed:
Fix test

loumalouomega committed:
Actually fix test

loumalouomega committed:
Extend Create

loumalouomega committed:
Extend create II

loumalouomega committed:
Extend create III

loumalouomega committed:
Extend create IV

loumalouomega committed:
Extend create V

loumalouomega committed:
Extend create VI

loumalouomega committed:
Extend create VII

loumalouomega committed:
Extend create VIII

loumalouomega committed:
Extend create IX

loumalouomega committed:
Extend create X

loumalouomega committed:
Merge branch 'master' into core/modelpart-create-geometries-methods

loumalouomega committed:
New methods

loumalouomega committed:
Testing

loumalouomega committed:
Exposing python

loumalouomega commented:
This is ready to merge, any further comment @tteschemacher ?

loumalouomega committed:
Merge branch 'master' into core/modelpart-create-geometries-methods

loumalouomega commented:
This is ready @KratosMultiphysics/technical-committee

adityaghantasala commented on kratos/geometries/geometry.h:
```diff
@@ -647,11 +647,92 @@ class Geometry
+     * @return Pointer to the new geometry
+     */
+    virtual Pointer Create(
+        const std::string& NewGeometryName,
```
```suggestion
        const std::string& rNewGeometryName,
```

loumalouomega replied:
Ok, But I cannot change it here, or I will break the code

adityaghantasala commented on kratos/geometries/geometry.h:
```diff
@@ -647,11 +647,92 @@ class Geometry
+     */
+    virtual Pointer Create(
+        const std::string& NewGeometryName,
+        PointsArrayType const& ThisPoints
```
```suggestion
        PointsArrayType const& rThisPoints
```

adityaghantasala commented on kratos/geometries/geometry.h:
```diff
@@ -647,11 +647,92 @@ class Geometry
+     * @return Pointer to the new geometry
+     */
+    virtual Pointer Create(
+        const std::string& NewGeometryName,
```
```suggestion
        const std::string& rNewGeometryName,
```

adityaghantasala commented on kratos/geometries/hexahedra_3d_20.h:
```diff
@@ -258,8 +258,25 @@ template<class TPointType> class Hexahedra3D20 : public Geometry<TPointType>
+    /// Constructor with Geometry Id
+    explicit Hexahedra3D20(
+        IndexType GeometryId,
+        const PointsArrayType& ThisPoints
```
```suggestion
        const PointsArrayType& rThisPoints
```

adityaghantasala commented on kratos/geometries/hexahedra_3d_20.h:
```diff
@@ -258,8 +258,25 @@ template<class TPointType> class Hexahedra3D20 : public Geometry<TPointType>
+    /// Constructor with Geometry Name
+    explicit Hexahedra3D20(
+        const std::string& GeometryName,
+        const PointsArrayType& ThisPoints
```
```suggestion
        const PointsArrayType& rThisPoints
```

adityaghantasala commented on kratos/geometries/hexahedra_3d_20.h:
```diff
@@ -347,17 +364,96 @@ template<class TPointType> class Hexahedra3D20 : public Geometry<TPointType>
-
-    typename BaseType::Pointer Create( PointsArrayType const& ThisPoints ) const override
+    typename BaseType::Pointer Create(
+        PointsArrayType const& ThisPoints
```
```suggestion
        PointsArrayType const& rThisPoints
```

adityaghantasala commented on kratos/geometries/hexahedra_3d_20.h:
```diff
@@ -347,17 +364,96 @@ template<class TPointType> class Hexahedra3D20 : public Geometry<TPointType>
+     */
+    typename BaseType::Pointer Create(
+        IndexType NewGeometryId,
+        PointsArrayType const& ThisPoints
```
```suggestion
        PointsArrayType const& rThisPoints
```

adityaghantasala commented on kratos/geometries/hexahedra_3d_20.h:
```diff
@@ -347,17 +364,96 @@ template<class TPointType> class Hexahedra3D20 : public Geometry<TPointType>
+     */
+    typename BaseType::Pointer Create(
+        const std::string& NewGeometryName,
+        PointsArrayType const& ThisPoints
```
```suggestion
        PointsArrayType const& rThisPoints
```

adityaghantasala commented on kratos/geometries/hexahedra_3d_27.h:
```diff
@@ -373,11 +391,90 @@ template<class TPointType> class Hexahedra3D27 : public Geometry<TPointType>
+     */
+    typename BaseType::Pointer Create(
+        IndexType NewId,  
+        PointsArrayType const& ThisPoints
```
```suggestion
        PointsArrayType const& rThisPoints
```

Do check for other places where this change is necessary.

loumalouomega committed:
Merge branch 'master' into core/modelpart-create-geometries-methods

adityaghantasala commented on kratos/sources/model_part.cpp:
```diff
@@ -1594,6 +1594,227 @@ void ModelPart::RemoveConditionsFromAllLevels(Flags IdentifierFlag)
 
+ModelPart::GeometryType::Pointer ModelPart::CreateNewGeometry(
+    std::string GeometryTypeName,
+    std::vector<IndexType> GeometryNodeIds
```
```suggestion
    const std::vector<IndexType> GeometryNodeIds
```

adityaghantasala commented on kratos/sources/model_part.cpp:
```diff
@@ -1594,6 +1594,227 @@ void ModelPart::RemoveConditionsFromAllLevels(Flags IdentifierFlag)
+ModelPart::GeometryType::Pointer ModelPart::CreateNewGeometry(
+    std::string GeometryTypeName,
+    IndexType GeometryId,
+    std::vector<IndexType> GeometryNodeIds
```
```suggestion
    const std::vector<IndexType> GeometryNodeIds
```

adityaghantasala commented on kratos/sources/model_part.cpp:
```diff
@@ -1594,6 +1594,227 @@ void ModelPart::RemoveConditionsFromAllLevels(Flags IdentifierFlag)
+}
+
+ModelPart::GeometryType::Pointer ModelPart::CreateNewGeometry(
+    std::string GeometryTypeName,
```
```suggestion
    const std::string GeometryTypeName,
```

adityaghantasala commented on kratos/sources/model_part.cpp:
```diff
@@ -1594,6 +1594,227 @@ void ModelPart::RemoveConditionsFromAllLevels(Flags IdentifierFlag)
+
+ModelPart::GeometryType::Pointer ModelPart::CreateNewGeometry(
+    std::string GeometryTypeName,
+    IndexType GeometryId,
```
```suggestion
    const IndexType GeometryId,
```

adityaghantasala commented on kratos/sources/model_part.cpp:
```diff
@@ -1594,6 +1594,227 @@ void ModelPart::RemoveConditionsFromAllLevels(Flags IdentifierFlag)
+ModelPart::GeometryType::Pointer ModelPart::CreateNewGeometry(
+    std::string GeometryTypeName,
+    const std::string& GeometryIdentifierName,
+    std::vector<IndexType> GeometryNodeIds
```
```suggestion
    const std::vector<IndexType> GeometryNodeIds
```
Do check other places and arguments where we can add `const`

loumalouomega committed:
Suggestion (I)

loumalouomega committed:
Suggestion (II)

loumalouomega committed:
Suggestion (III)

loumalouomega committed:
Suggestion (IV)

loumalouomega committed:
Suggestion (V)

loumalouomega committed:
Suggestion (VI)

loumalouomega committed:
Suggestion (VII)

loumalouomega committed:
Suggestion (VIII)

loumalouomega committed:
Suggestion (IX)

loumalouomega committed:
Suggestion (X)

loumalouomega committed:
Merge branch 'master' into core/modelpart-create-geometries-methods

pablobecker commented on kratos/sources/model_part.cpp:
```diff
@@ -1594,6 +1594,227 @@ void ModelPart::RemoveConditionsFromAllLevels(Flags IdentifierFlag)
+    KRATOS_TRY
+
+        if (IsSubModelPart()) {
+            GeometryType::Pointer p_new_geometry = mpParentModelPart->CreateNewGeometry(GeometryTypeName, pGeometryNodes);
```
does this also work if there are multiple (several!) levels of submodelparts?

loumalouomega replied:
Should be, it is the same strategy followed in the elements creation

pablobecker replied:
I'll have to check. it seems strange that if adding it in the current one is not enough, doing it from the parent one would be enough..

in the same line, there's a bug i detected and still have to fix in the remeshing of the meshing app, that only works if you have a single level of submodelparts.

pablobecker replied:
oh, it's recursive, allright then

loumalouomega commented:
Anyone?

philbucher commented:
this is still under discussion in @KratosMultiphysics/technical-committee

loumalouomega committed:
Merge branch 'master' into core/modelpart-create-geometries-methods

loumalouomega committed:
Some corrections

loumalouomega committed:
Revert "Simplify"

This reverts commit efefc140ea0820e2e1eb2e774a9a0154026674bf.

loumalouomega committed:
Simplify

loumalouomega committed:
Missing const

loumalouomega committed:
Simplify base methods

loumalouomega committed:
Clean up duplicated in Hexahedra

loumalouomega committed:
Clean up duplicated in lines

loumalouomega committed:
Clean up duplicated in points

loumalouomega committed:
Clean up duplicated in prism

loumalouomega committed:
Clean up duplicated in quad

loumalouomega committed:
Clean up duplicated in tetrahedra

loumalouomega committed:
Clean up duplicated in triangle

loumalouomega committed:
Clean up duplicated in sphere

loumalouomega committed:
Clean up duplicated in qpoints

loumalouomega committed:
Merge branch 'master' into Core/modelpart-create-geometries-methods

loumalouomega committed:
Revert "Separate"

This reverts commit eedfe100fcd9d6c9de1d65602638e41b88f10e73.

loumalouomega committed:
Separate

loumalouomega committed:
Changes

loumalouomega committed:
Correct Id generation

loumalouomega committed:
Minor

loumalouomega committed:
Fix

loumalouomega committed:
Final fix

pooyan-dadvand commented on kratos/geometries/geometry.h:
```diff
@@ -647,9 +647,120 @@ class Geometry
     {
-        return Pointer( new Geometry( ThisPoints, mpGeometryData ) );
+        // Create geometry
+        auto p_geom = this->Create(0, pGeometry);
```
I would add a debug check if pGeometry is not null

pooyan-dadvand replied:
BTW, what would be the use case for this? Shouldn't be `Clone`?

I mean, if we need a clone of existing geometry, so let's call it `Clone`. If we only need the array of points then passing geometry to the `Create(PointsArrayType)` would work because  geometry is already an array of points.

So should it be a `Clone` or there is another use case justifying this one?

loumalouomega replied:
I chose Create because I thought it was the proper name, but can be renamed as Clone if necessary

pooyan-dadvand replied:
But do we need it? Please note that then we should have to define it in all derived geometries.

pooyan-dadvand replied:
I found an answer to my concern. It should be Create as the id is regenerated for the new geometry. (Clone would maintain the id)

pooyan-dadvand replied:
And the difference to the PointsArrayType is the data container that is also copied.

pooyan-dadvand commented on kratos/geometries/geometry.h:
```diff
@@ -647,9 +647,120 @@ class Geometry
+     * @param rThisPoints the nodes of the new geometry
+     * @return Pointer to the new geometry
+     */
+    virtual Pointer Create(
```
I think that we don't need to have this one virtual. In fact, it would be better to centralize the automatic creation of the Id in the geometry base class.

pooyan-dadvand replied:
So why you keep this virtual?

pooyan-dadvand commented on kratos/geometries/geometry.h:
```diff
@@ -647,9 +647,120 @@ class Geometry
+     * @param rThisPoints the nodes of the new geometry
+     * @return Pointer to the new geometry
+     */
+    virtual Pointer Create(
```
This should be the only virtual one.

pooyan-dadvand commented on kratos/geometries/geometry.h:
```diff
@@ -647,9 +647,120 @@ class Geometry
+     * @param rThisPoints the nodes of the new geometry
+     * @return Pointer to the new geometry
+     */
+    virtual Pointer Create(
```
Like the other one, I would not making this one virtual. (we may create the id and use the other one)

loumalouomega replied:
I need to study the virtuals, I am using several of them

pooyan-dadvand replied:
I would not make this virtual unless we find a use case for it. Let keep the only version here and derived class only to override the Create with Id which is called here.

pooyan-dadvand reviewed (commented):
Sorry guys but I forgot the use case for the `Create(Geometry::Pointer)` methods. Whould you please explain it again.

loumalouomega committed:
Merge branch 'master' into core/modelpart-create-geometries-methods

loumalouomega committed:
Check null pointer

loumalouomega committed:
Minor

philbucher committed:
Merge branch 'master' into core/modelpart-create-geometries-methods

pooyan-dadvand reviewed (commented):
Ready for @KratosMultiphysics/technical-committee

pooyan-dadvand commented on kratos/geometries/geometry.h:
```diff
@@ -647,9 +647,121 @@ class Geometry
+     * @return Pointer to the new geometry
+     */
+    virtual Pointer Create(
+        Pointer pGeometry
```
Please change the Pointer to a `const Geometry&  rGeometry` to avoid the performance loss due to the smart pointer reference counting

loumalouomega replied:
Okay

pooyan-dadvand reviewed (commented):
@KratosMultiphysics/technical-committee approves the additional interfaces.

Nevertheless, there are still minor implementation details to be finished. (Omitting some virtuals and passing by reference)

loumalouomega committed:
Merge branch 'master' into core/modelpart-create-geometries-methods

loumalouomega committed:
Update geometries

loumalouomega committed:
Update modelpart

pooyan-dadvand reviewed (commented):
I would only keep virtual the ones that you are overriding in the driven classes.

loumalouomega committed:
Merge branch 'master' into core/modelpart-create-geometries-methods

loumalouomega committed:
Remove Clone methods (98% commented)

loumalouomega committed:
Remove unused virtual

pooyan-dadvand reviewed (approved):
Thanks @loumalouomega! :+1:
